src/snapshot.zig

```diff
@@ -0,0 +1,159 @@
+const std = @import("std");
+const base = @import("base.zig");
+const parse = @import("check/parse.zig");
+
+var verbose_log: bool = false;
+
+fn log(comptime fmt: []const u8, args: anytype) void {
+    if (verbose_log) {
+        std.log.info(fmt, args);
+    }
+}
+
+pub fn main() !void {
+    var gpa_impl = std.heap.GeneralPurposeAllocator(.{}){};
+    defer {
+        _ = gpa_impl.deinit();
+    }
+    const gpa = gpa_impl.allocator();
+
+    const args = try std.process.argsAlloc(gpa);
+    defer std.process.argsFree(gpa, args);
+
+    // Check for verbose flag
+    var target_snapshot: ?[]const u8 = null;
+
+    for (args[1..]) |arg| {
+        if (std.mem.eql(u8, arg, "--verbose")) {
+            verbose_log = true;
+        } else if (target_snapshot == null) {
+            target_snapshot = arg;
+        }
+    }
+
+    const snapshots_dir = "src/snapshots";
+    var file_count: usize = 0;
+    var timer = std.time.Timer.start() catch unreachable;
+
+    if (target_snapshot != null) {
+        // Only for the specified snapshot file
+        try processSnapshotFile(target_snapshot.?, gpa);
+        file_count += 1;
+    } else {
+        // For each of the files in the directory
+        var dir = try std.fs.cwd().openDir(snapshots_dir, .{ .iterate = true });
+        defer dir.close();
+
+        var dir_iterator = dir.iterate();
+        while (try dir_iterator.next()) |entry| {
+            if (entry.kind != .file) continue;
+            if (!std.mem.endsWith(u8, entry.name, ".txt")) continue;
+
+            const full_path = try std.fmt.allocPrint(gpa, "{s}/{s}", .{ snapshots_dir, entry.name });
+            defer gpa.free(full_path);
+
+            try processSnapshotFile(full_path, gpa);
+
+            file_count += 1;
+        }
+    }
+
+    const duration_ms = timer.read() / std.time.ns_per_ms;
+
+    // Print a summary
+    std.log.info("processed {d} snapshots in {d}ms.", .{ file_count, duration_ms });
+}
+
+/// Represents the different sections of a snapshot file.
+const Section = enum {
+    None,
+    Meta,
+    Source,
+    Parse,
+};
+
+/// Stores the contents for a snapshot to be written to file
+const SnapshotContents = struct {
+    meta: std.ArrayList(u8),
+    source: std.ArrayList(u8),
+    parse: std.ArrayList(u8),
+
+    fn init(allocator: std.mem.Allocator) SnapshotContents {
+        return .{
+            .meta = std.ArrayList(u8).init(allocator),
+            .source = std.ArrayList(u8).init(allocator),
+            .parse = std.ArrayList(u8).init(allocator),
```

Comment:
nit: allocators should be named `gpa` if they are meant to be general and not arenas. Which is the case for anything that uses arraylists.
```suggestion
    fn init(gpa: std.mem.Allocator) SnapshotContents {
        return .{
            .meta = std.ArrayList(u8).init(gpa),
            .source = std.ArrayList(u8).init(gpa),
            .parse = std.ArrayList(u8).init(gpa),
```

Reply:
Done in f29dcf18ea